### ISSUE TITLE
HTMLArea causing internal server errors on editItem()

### ISSUE THREAD

**Issue by jmartens** [pradosoft/prado#710]
I am porting a Prado 3 application to Prado 4 and most off it went sweetly, but there is an edit field with a HTMLArea which seems to generate internal server errors when I try to go to edit mode.

From the chrome debugger log this is what I found:
```
ajax3.js:71 Prado\Exceptions\TPhpErrorException with message "[Notice] Trying to access array offset on value of type null (@line 190 in file xxx\prado4\pradosoft\prado\framework\Web\UI\ActiveControls\TCallbackPageStateTracker.php)." in ${PradoFramework}\Web\UI\ActiveControls\TCallbackPageStateTracker.php(190)
ajax3.js:72 Stack trace:
ajax3.js:80 #0 ${PradoFramework}\Web\UI\ActiveControls\TCallbackPageStateTracker.php(190)
ajax3.js:90 Uncaught TypeError: Cannot read property 'join' of undefined
    at Object.logFormatException (ajax3.js:90)
    at klass.errorHandler (ajax3.js:419)
    at fire (jquery.js:3291)
    at Object.fireWith [as rejectWith] (jquery.js:3421)
    at done (jquery.js:9535)
    at XMLHttpRequest.<anonymous> (jquery.js:9785)
```

This seems to be caused by updateStyle() referencing $style which is set to null. When I remove the contents of this function the page switches to edit mode and the HTMLArea renders as it did in the Prado 3 version.

Is there something I am unaware of and have to modify or am I running into some kind of bug?

**Comment by ctrlaltca** [comment]
Hi, and thank you for the report.
I understand where the problem occurs and why, but i can't really reproduce the issue on a `<THtmlArea>` or a `<TActiveHtmlArea>` control.
An active control triggered an ajax callback, Prado computed the style differences between the old and the new state of the htmlarea, and is now pushing the differences to the clientside.
The variable containing these differences should be either an empty `stdClass` (meaning no difference is present) or an array of items. I see no evident way in the code on how it can become `null`.
It seems you also found another bug on the console logger:
`ajax3.js:90 Uncaught TypeError: Cannot read property 'join' of undefined`
It looks like the exception description sent by Prado to the browser is missing some pieces.
Can I ask you for more details about this?
Having a minimal testcase for reproducing the bug would be awesome: you can use the [HtmlArea demo](https://github.com/pradosoft/prado-demos/tree/master/quickstart/protected/pages/Controls/Samples/THtmlArea) as a base for the testcase.
If you can't create the testcase, can you at least attach to this ticket a copy of the raw network reply sent by Prado when the error occurs (removing any privacy-sensitive content, of course)?
Once we get the console logger fixed, weshould be able to get a full stacktrace of the main problem.
Thank you

**Comment by jmartens** [comment]
I tried to setup a minimal example for you to try and test with which fails if `updateStyle(style)` looks like [this](https://github.com/pradosoft/prado/blob/d07b96f82d77d23b8a68dcc7aeda528ec209707b/framework/Web/UI/ActiveControls/TCallbackPageStateTracker.php#L188) and works if I comment out that code block like so:
```php
	protected function updateStyle($style)
	{
        /*
         * Commented out to enable edit functionality on HTMLArea @see https://github.com/pradosoft/prado/issues/710
        if ($style['CssClass'] !== null) {
            $this->client()->setAttribute($this->_control, 'class', $style['CssClass']);
        }
        if (is_array($style['Style']) && count($style['Style']) > 0) {
            $this->client()->setStyle($this->_control, $style['Style']);
        }*/
	}
```

Below is a minimal example (requiring a template and `com:TContentPlaceHolder` with Id `col3_content`

test.tpl
```Smarty
<com:TContent
	Id="col3_content">
	<com:TPanel
		Id="PanelTodo">
		<h2><%[Todo]%></h2>
		<com:TActiveDataGrid
			AllowPaging="true"
            AllowSorting="true"
            ItemStyle.CssClass="item even"
            AlternatingItemStyle.CssClass="item odd"
            EditItemStyle.CssClass="item edit"
            HeaderStyle.CssClass="item header"
			AutoGenerateColumns="false"
			CssClass="grid"
			DataKeyField="Id"
			DataSource="EV_PatientRecord"
			Id="Todo"
			EmptyTemplate="<%[ No results ]%>"
			onCancelCommand="cancelTodoItem"
			onEditCommand="editTodoItem"
			onUpdateCommand="updateTodoItem"
			onPageIndexChanged="changePage"
			onSortCommand="sortDataGrid">
        <com:TTemplateColumn
            HeaderText="<%[Patient]%>"
            Id="PatientIdColumn" >
            <prop:ItemTemplate>
                <com:THyperLink
                    NavigateUrl="/reports/<%# $this->Parent->Data['PatientId'] %>"
                    Text="<%# $this->Parent->Data['PatientId'] %>"
                    ToolTip="<%# Prado::localize('View reports for patient %s', array('patient' => $this->Parent->Data['PatientId'])) %>" />
            </prop:ItemTemplate>
            <prop:EditItemTemplate></prop:EditItemTemplate>
        </com:TTemplateColumn>
        <com:TTemplateColumn
            HeaderText="<%[Plan]%>"
            Id="PlanIdColumn" >
            <prop:ItemTemplate>
                <com:THyperLink
                    NavigateUrl="/reports/<%# $this->Parent->Data['PatientId'] %>/<%# $this->Parent->Data['PlanId'] %>"
                    Text="<%# $this->Parent->Data['PlanId'] %>"
                    ToolTip="<%# Prado::localize('View reports for patient %s and plan %s', array('patient' => $this->Parent->Data['PatientId'], 'plan' => $this->Parent->Data['PlanId'])) %>" />
            </prop:ItemTemplate>
            <prop:EditItemTemplate></prop:EditItemTemplate>
        </com:TTemplateColumn>
        <com:TActiveBoundColumn
            Id="MeasurementTypeColumn"
            HeaderText="<%[Type]%>"
            DataField="MeasurementType"
            ReadOnly="true"
            SortExpression="MeasurementType" />
        <com:TActiveBoundColumn
            Id="FractionNumberColumn"
            HeaderText="<%[Fraction]%>"
            DataField="FractionNumber"
            ReadOnly="true" />
        <com:TActiveCheckBoxColumn
            Id="Pass2DColumn"
            HeaderText="<%[Pass 2D]%>"
            DataField="Pass2D"
            ReadOnly="true"
            SortExpression="Pass2D" />
        <com:TActiveCheckBoxColumn
            Id="Pass3DColumn"
            HeaderText="<%[Pass 3D]%>"
            DataField="Pass3D"
            ReadOnly="true"
            SortExpression="Pass3D" />
        <com:TActiveDropDownListColumn
            ID="CommentTypeColumn"
            HeaderText="<%[CommentId]%>"
            DataValueField="CommentId">
            <com:TListItem Value="1" Text="<%[COMMENT_TYPE1]%>" />
            <com:TListItem Value="2" Text="<%[COMMENT_TYPE2]%>" />
            <com:TListItem Value="3" Text="<%[COMMENT_TYPE3]%>" />
            <com:TListItem Value="4" Text="<%[COMMENT_TYPE4]%>" />
            <com:TListItem Value="5" Text="<%[COMMENT_TYPE5]%>" />
            <com:TListItem Value="6" Text="<%[COMMENT_TYPE6]%>" />
            <com:TListItem Value="7" Text="<%[COMMENT_TYPE7]%>" />
            <com:TListItem Value="8" Text="<%[COMMENT_TYPE8]%>" />
            <com:TListItem Value="9" Text="<%[COMMENT_TYPE9]%>" />
            <com:TListItem Value="11" Text="<%[COMMENT_TYPE11]%>" />
            <com:TListItem Value="10" Text="<%[COMMENT_TYPE10]%>" />
        </com:TActiveDropDownListColumn>
        <com:TTemplateColumn
            Id="CommentColumn">
            <prop:HeaderTemplate><%[Comment]%></prop:HeaderTemplate>
            <prop:ItemTemplate />
            <prop:EditItemTemplate>
                <com:THtmlArea
                    Id="Comment"
                    AutoTrim="true"
                    Text="<%#$this->Parent->Data['Comment']%>">
                    <prop:Options>
                        plugins: "fullscreen,spellchecker,searchreplace",
                        theme : "advanced",
                        theme_advanced_buttons1 : "search,replace,|,fullscreen,|,spellchecker,insertdate,inserttime,|,cut,copy,paste,pastetext,|,undo,redo,|,bullist,numlist,|,outdent,indent,blockquote,|,link,unlink,|,code,cleanup,|,help",
                        theme_advanced_buttons2 : "bold,italic,underline,strikethrough,|,forecolor,backcolor,|,sub,sup,|,charmap,|,hr,removeformat,visualaid",
                        theme_advanced_resize_horizontal : false,
                        theme_advanced_resizing : true
                    </prop:Options>
                </com:THtmlArea>
            </prop:EditItemTemplate>
    	</com:TTemplateColumn>
	    <com:TActiveEditCommandColumn
            ButtonType="ImageButton"
 	    	CancelImageUrl="./images/16px-Gnome-process-stop.png"
            CancelText="<%[Cancel]%>"
 	    	EditImageUrl="./images/16px-Gnome-accessories-text-editor.png"
            EditText="<%[Edit]%>"
            HeaderText="<%[Edit]%>"
 	    	UpdateImageUrl="./images/16px-Gnome-media-floppy.png"
            UpdateText="<%[Save]%>" />
		</com:TActiveDataGrid>
	</com:TPanel>
	<com:TPanel
		Id="PanelHistory">
		<h2><%[History]%></h2>
		<com:TActiveDataGrid
			AllowPaging="true"
			AllowSorting="false"
			CssClass="grid"
            ItemStyle.CssClass="item even"
            AlternatingItemStyle.CssClass="item odd"
            EditItemStyle.CssClass="item edit"
            HeaderStyle.CssClass="item header"
			AutoGenerateColumns="false"
			Id="History"
			EmptyTemplate="<%[ No results ]%>"
			onPageIndexChanged="changePage">
			<com:TTemplateColumn
				HeaderText="<%[Patient]%>" >
				<prop:ItemTemplate>
					<com:THyperLink
						NavigateUrl="/reports/<%# $this->Parent->Data['PatientId'] %>"
						Text="<%# $this->Parent->Data['PatientId'] %>"
						ToolTip="<%# Prado::localize('View reports for patient %s', array('patient' => $this->Parent->Data['PatientId'])) %>" />
				</prop:ItemTemplate>
				<prop:EditItemTemplate></prop:EditItemTemplate>
			</com:TTemplateColumn>
			<com:TTemplateColumn
				HeaderText="<%[Plan]%>" >
				<prop:ItemTemplate>
					<com:THyperLink
						NavigateUrl="/reports/<%# $this->Parent->Data['PatientId'] %>/<%# $this->Parent->Data['PlanId'] %>"
						Text="<%# $this->Parent->Data['PlanId'] %>"
						ToolTip="<%# Prado::localize('View reports for patient %s and plan %s', array('patient' => $this->Parent->Data['PatientId'], 'plan' => $this->Parent->Data['PlanId'])) %>" />
				</prop:ItemTemplate>
				<prop:EditItemTemplate></prop:EditItemTemplate>
			</com:TTemplateColumn>
			<com:TActiveBoundColumn
				HeaderText="<%[Type]%>"
				DataField="MeasurementType" />
			<com:TActiveBoundColumn
				HeaderText="<%[Fraction]%>"
				DataField="FractionNumber" />
			<com:TActiveCheckBoxColumn
				HeaderText="<%[Pass 2D]%>"
				DataField="Pass2D" />
			<com:TActiveCheckBoxColumn
				HeaderText="<%[Pass 3D]%>"
				DataField="Pass3D" />
			<com:TActiveTemplateColumn
				HeaderText="<%[Comment ID]%>">
				<prop:itemTemplate><%# Prado::localize(sprintf('COMMENT_TYPE%s', $this->NamingContainer->Data['Comment']Id)) %></prop:itemTemplate>
			</com:TActiveTemplateColumn >
			<com:TActiveBoundColumn
				HeaderText="<%[Comment by]%>"
				DataField="CommentUser" />
			<com:TActiveBoundColumn
				HeaderText="<%[Comment]%>"
				DataField="Comment" />
		</com:TActiveDataGrid>
	</com:TPanel>
</com:TContent>
```
test.php
```php
<?php

class test extends TPage {

	private $_todo;

	protected function loadData() {

        $this->_todo = [array(
            'Id' => "1",
            'PatientId' => "PatientID",
            'PlanId' => "PlanID",
            'MeasurementType' => "MeasurementType",
            'FractionNumber' => "1",
            'Pass2D' => "1",
            'Pass3D' => "0",
            'CommentId' => null,
            'Comment' => null,
            'CommentUser' => null,
            'NewIssue' => "1",
            'TimeStamp' => "2020-03-26 17:35:09.040"
        )];
	}

	protected function getTodo() {

			if($this->_todo === null)
				$this->loadData();
			return $this->_todo;

	}

	public function cancelTodoItem($sender, $param) {

		$this->Todo->EditItemIndex=-1;
		$this->Todo->DataSource=$this->getTodo();
		$this->Todo->dataBind();

	}

	public function editTodoItem($sender, $param) {

		$this->Todo->EditItemIndex=$param->Item->ItemIndex;
		$this->Todo->DataSource=$this->getTodo();
		$this->Todo->dataBind();

	}

	public function updateTodoItem($sender, $param) {

        // Functional code removed for condensed example
            
		// Reload the form
		$this->Todo->EditItemIndex=-1;
		$this->Todo->DataSource=$this->getTodo();
		$this->dataBind();

	}

	public function onLoad($param) {

		parent::onLoad($param);

		if(!$this->IsPostBack && !$this->IsCallBack)
		{           
			$this->Todo->DataSource = $this->getTodo();
			$this->dataBind();
		}

	}

}

?>
```

**Comment by jmartens** [comment]
> It seems you also found another bug on the console logger:
> `ajax3.js:90 Uncaught TypeError: Cannot read property 'join' of undefined`
> It looks like the exception description sent by Prado to the browser is missing some pieces.
> Can I ask you for more details about this?

I saved the debugger output of the console to a log. I will see if I can attach that.
[localhost-1585297547909.log](https://github.com/pradosoft/prado/files/4393191/localhost-1585297547909.log)


> If you can't create the testcase, can you at least attach to this ticket a copy of the raw network reply sent by Prado when the error occurs (removing any privacy-sensitive content, of course)?

Does the above mentioned attachment do or do you need more? How am I to catch the raw network reply? Do you have any instructions for that?

**Comment by ctrlaltca** [comment]
I've ran your testcase under Prado 4.1.0, and I've been able to reproduce the error.
I've also ran the testcase using dev-master, and I have a good news since it seems the problem has already been solved in https://github.com/pradosoft/prado/commit/b31e08f0655a773a3d58e15462ade263ce3414c4.
I'll tag a new release 4.1.1 right now including the fix.
Thank you a lot for providing the testcase.

**Comment by ctrlaltca** [comment]
Prado 4.1.1 release, can you try to update and see if it fixes the your problem?

**Comment by jmartens** [comment]
> Prado 4.1.1 release, can you try to update and see if it fixes the your problem?

Yes that fixed the issue with the THtmlArea not rendering. Thanks for the quick reponse and turnover.

Do you need a seperate issue for the other problem (below) discovered with the ajax request?

> It seems you also found another bug on the console logger:
> `ajax3.js:90 Uncaught TypeError: Cannot read property 'join' of undefined`
> It looks like the exception description sent by Prado to the browser is missing some pieces.
> Can I ask you for more details about this?

**Comment by ctrlaltca** [comment]
> Do you need a seperate issue for the other problem (below) discovered with the ajax request?

There's no need for a new issue, thank you.

**Comment by ctrlaltca** [comment]
Hi, I'l like you to know that I just reworked the previous fix for this issue in 7cb178067c5b55eaf504c35f7a0f0ae1ffd0f473
The old fix could cause an object to lose its css class when its style was updated during a callback.
I've already re-checked the testcase you provided against the new fix.
